### PR TITLE
Implemented PEC, PMC, Bloch Boundary Conditions

### PR DESCRIPTION
**Pull request by bruxillensis**
Implemented #40. Summary of changes:
1. Added perfect electric, perfect magnetic, and bloch BC's
2. Made periodic BC an alias of bloch with k=0 to reduce redundancy, removed periodic.py
3. Refactoring of field padding, field reset, and E/H updates to reduce redundancy and move logic into boundary classes where possible.
4. Added simulation tests for all BC's
5. Added 1D bloch simulation example
band_diagram

---
## Summary by cubic
Adds `BlochBoundary` (with phase vector), `PerfectElectricConductor`, and `PerfectMagneticConductor`. Periodic is now Bloch with k=0. Implements Linear #40, moves padding and field resets into boundary classes, and adds a 1D Bloch band-structure example.

- **New Features**
  - `BlochBoundary` with `bloch_vector`; complex fields auto-enabled via `SimulationConfig.use_complex_fields` (None=auto, True/False override).
  - `PerfectElectricConductor` and `PerfectMagneticConductor`; zero tangential E/H after each update.
  - `examples/bloch_band_structure.py` for 1D photonic crystal bands.
  - Physics tests for Bloch (incl. oblique), PEC/PMC reflection, mixed boundaries, symmetry; plus time-reversal reconstruction and gradient tests; restored periodic cases.
  - Poynting flux detector returns the real part; added validation to block complex→real dtype downcasts in `DtypeConversion`.

- **Migration**
  - Replace `PeriodicBoundary` with `BlochBoundary` using `bloch_vector=(0, 0, 0)`; `boundary_type="periodic"` in config still works.
  - `interpolate_fields` now takes padded arrays; call `pad_fields` first (the `periodic_axes` arg was removed).
  - Custom boundaries should use `BaseBoundary.uses_wrap_padding`, `apply_pad_correction`, and `apply_field_reset`.
  - For non-zero Bloch k, ensure complex fields are enabled (auto or set `use_complex_fields=True`).

<sup>Written for commit 3f84b4fc2675971a5fac604a92f138b6a9f72c0e. Summary will update on new commits.</sup>

### CODE REVIEW

**Comment by ymahlau** [comment]
@cubic-dev-ai review

File: src/fdtdx/core/physics/curl.py [hunk 1]
```diff
@@ -25,9 +24,16 @@ def interpolate_fields(
...
+        E_pad: Pre-padded electric field array of shape (3, Nx+2, Ny+2, Nz+2)
+        H_pad: Pre-padded magnetic field array of shape (3, Nx+2, Ny+2, Nz+2)
+
+    Returns:
+        Tuple of (E_interp, H_interp), each of shape (3, Nx, Ny, Nz)
     """
-    E_field = pad_fields(E_field, periodic_axes)
-    H_field = pad_fields(H_field, periodic_axes)
+    E_field = E_pad
+    H_field = H_pad
```
Comment (ymahlau) [inline src/fdtdx/core/physics/curl.py]: This renaming can be removed now

File: src/fdtdx/fdtd/initialization.py [hunk 2]
```diff
@@ -32,7 +33,7 @@
 
 
 def place_objects(
-    object_list: Sequence[SimulationObject],
+    object_list: list[SimulationObject],
```
Comment (ymahlau) [inline src/fdtdx/fdtd/initialization.py]: This should be kept a list for compatibility and user typing

File: src/fdtdx/fdtd/container.py [hunk 3]
```diff
@@ -77,12 +79,24 @@ def pml_objects(self) -> list[PerfectlyMatchedLayer]:
...
+    def pec_objects(self) -> list[PerfectElectricConductor]:
+        return [o for o in self.objects if isinstance(o, PerfectElectricConductor)]
+
+    @property
+    def pmc_objects(self) -> list[PerfectMagneticConductor]:
+        return [o for o in self.objects if isinstance(o, PerfectMagneticConductor)]
+
+    @property
+    def bloch_objects(self) -> list[BlochBoundary]:
+        return [o for o in self.objects if isinstance(o, BlochBoundary)]
```
Comment (ymahlau) [inline src/fdtdx/fdtd/container.py:95]: For this PR this is well done, but I created a github issue that this filtering needs to be refactored (#266)

File: src/fdtdx/objects/boundaries/bloch.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    Generalizes periodic boundary conditions with a phase shift:
+        F(x + L) = F(x) * exp(i * k_bloch * L)
+
+    When the Bloch vector is zero, this is equivalent to a standard periodic
+    boundary. Complex-valued field arrays are only required when the Bloch
+    vector has non-zero components.
+    """
+
+    #: Bloch wave vector components (k_x, k_y, k_z) in units of rad/m.
+    bloch_vector: tuple[float, float, float] = frozen_field(default=(0.0, 0.0, 0.0))
```
Comment (ymahlau) [inline src/fdtdx/objects/boundaries/bloch.py:26]: If i see correctly, only one component of the bloch vector is used, right? Then I think it does not make sense to force the user to specify all three components.

File: tests/unit/fdtd/test_backward.py [hunk 5]
```diff
@@ -152,119 +155,26 @@ def test_pml_regions_zeroed(self, mock_arrays, mock_objects, key, patched_update
...
+            reset_fields=True,
+            fields_to_reset=("E",),
+        )
 
-        assert "E" in aset_log
-        assert "H" in aset_log
-        # The destination region (y=8:10) should be overwritten with values
-        # from y=9:10, so both rows should equal the original y=9 values (3.0).
-        e_result = aset_log["E"]
-        assert jnp.allclose(e_result[:, :, 8:10, :], e_result[:, :, 9:10, :])
```
Comment (ymahlau) [inline tests/unit/fdtd/test_backward.py:267]: Is there a reason that these test cases were deleted?

**Comment by ymahlau** [comment]
FYI, I have removed the changelog.md since it lead to too many unnecessary merge conflicts. I was using it to generate release notes, but this can be done automatically from git commits and PR

File: src/fdtdx/objects/boundaries/bloch.py (hunk 4 above)
Comment (bruxillensis) [reply src/fdtdx/objects/boundaries/bloch.py:26]: The full vector is stored since the bloch vector can be 3D, and the phase shift that gets applied depends on the axis the individual bloch boundaries are on. So for example can be bloch_vector=(kx, ky, 0) for a 2D photonic crystal simulation, full vector is stored in each boundary, but the kx phase is applied to the x-axis boundaries, the ky phase is applied to the y-axis boundaries.

File: tests/unit/fdtd/test_backward.py (hunk 5 above)
Comment (bruxillensis) [reply tests/unit/fdtd/test_backward.py:267]: Sorry, I missed adding these into the apply_field_reset tests when refactoring. I've added them back into TestPeriodicApplyFieldReset.

**Comment by bruxillensis** [comment]
Hey @ymahlau I think I have addressed the issues! Let me know what you think.

**Comment by ymahlau** [comment]
Looks good to me! Many thanks!
